packages/core/src/browser/shell/tab-bar-decorator.ts

```diff
@@ -0,0 +1,175 @@
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, named, postConstruct } from 'inversify';
+import { Event, Emitter, Disposable, DisposableCollection, ContributionProvider } from '../../common';
+import { Title, Widget } from '@phosphor/widgets';
+
+export const TabBarDecorator = Symbol('TabBarDecorator');
+
+export interface TabBarDecorator {
+
+    /**
+     * The unique identifier of the tab bar decorator.
+     */
+    readonly id: string;
+
+    readonly onDidChangeDecorations: Event<(titles: Title<Widget>[]) => void>;
+
+    /**
+     * Decorate tabs by the associated editor URI.
+     * @returns A map from URI of the tab to its decoration data.
+     */
+    decorate(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data>;
+}
+
+@injectable()
+export class TabBarDecoratorService implements Disposable {
+
+    protected readonly onDidChangeDecorationsEmitter = new Emitter<void>();
+
+    readonly onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
+
+    protected readonly toDispose = new DisposableCollection();
+
+    @inject(ContributionProvider) @named(TabBarDecorator)
+    protected readonly contributions: ContributionProvider<TabBarDecorator>;
+
+    protected decorators: ReadonlyArray<TabBarDecorator>;
+
+    @postConstruct()
+    protected init(): void {
+        this.decorators = this.contributions.getContributions();
+        this.toDispose.pushAll(this.decorators.map(decorator =>
+            decorator.onDidChangeDecorations(data =>
+                this.onDidChangeDecorationsEmitter.fire(undefined)
+            ))
+        );
+
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    getDecorations(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data[]> {
+        const changes: Map<string, TabBarDecoration.Data[]> = new Map();
+        for (const decorator of this.decorators) {
+            console.log(decorator);
+            for (const [id, data] of (decorator.decorate(titles)).entries()) {
+                if (changes.has(id)) {
+                    changes.get(id)!.push(data);
+                } else {
+                    changes.set(id, [data]);
+                }
+            }
+        }
+        return changes;
+    }
+}
+
+export namespace TabBarDecoration {
+
+    export namespace Styles {
```

Comment:
There is a lot of duplication between the `TabBarDecoration` and `TreeDecoration`.
Can we simplify the code and refactor common code under a base `Decoration` that these two decoration can extend?

Reply:
I've been thinking about this as well. Sure I will do it ;)